eclair-core/src/main/scala/fr/acinq/eclair/payment/offer/OfferManager.scala

```diff
@@ -0,0 +1,331 @@
+   */
+  sealed trait HandlerCommand
+
+  /**
+   * When an invoice request is received, a `HandleInvoiceRequest` is sent to the handler registered for this offer.
+   * @param replyTo        The handler must reply with either `InvoiceRequestActor.ApproveRequest` or `InvoiceRequestActor.RejectRequest`.
+   * @param invoiceRequest The invoice request to accept or reject. It is guaranteed to be valid for the offer.
+   */
+  case class HandleInvoiceRequest(replyTo: ActorRef[InvoiceRequestActor.Command], invoiceRequest: InvoiceRequest) extends HandlerCommand
+
+  /**
+   * When a payment is received for an offer invoice, a `HandlePayment` is sent to the handler registered for this offer.
+   * The handler may receive several `HandlePayment` for the same payment, usually because of multi-part payments.
+   * @param replyTo The handler must reply with either `PaymentActor.ApprovePayment` or `PaymentActor.RejectPayment`.
+   * @param offerId The id of the offer in case a single handler handles multiple offers.
+   * @param data    Data provided by the handler when generating the invoice, for its own use.
+   */
+  case class HandlePayment(replyTo: ActorRef[PaymentActor.Command], offerId: ByteVector32, data: ByteVector) extends HandlerCommand
+
+  case class RegisteredOffer(offer: Offer, nodeKey: PrivateKey, pathId_opt: Option[ByteVector32], handler: ActorRef[HandlerCommand])
+
+  case class PaymentMetadata(preimage: ByteVector32,
+                             payerKey: PublicKey,
+                             createdAt: TimestampSecond,
+                             quantity: Long,
+                             amount: MilliSatoshi,
+                             extraData: ByteVector)
+
+  private val metadataCodec: Codec[PaymentMetadata] =
+    (("preimage" | bytes32) ::
+      ("payerKey" | publicKey) ::
+      ("createdAt" | timestampSecond) ::
+      ("quantity" | uint64overflow) ::
+      ("amount" | millisatoshi) ::
+      ("extraData" | bytes)).as[PaymentMetadata]
+
+  /**
+   * Metadata that will be included in the blinded route so that we can recover it when the payment happens without
+   * needing to store it in a database (which would be a DoS vector).
+   * The data is signed so that it can't be forged by the payer, it is also encrypted as part of the blinding route so
+   * the payer can't read it.
+   *
+   * It contains
+   * - the offer id
+   * - the preimage of the payment hash
+   * - the payer key
+   * - the creation time of the invoice
+   * - the quantity of items bought
+   * - the amount to pay
+   * - optional data from the handler
+   *
+   * This data is what we need to recreate an invoice similar to the one that was actually sent to the payer. It will
+   * not be exactly the same (notably the blinding route will be different) but it will contain what we need to fulfill
+   * the payment HTLC.
+   *
+   * It takes 181 bytes plus what the handler adds.
+   */
+  case class SignedMetadata(signature: ByteVector64,
```

Comment:
Using a signature is quite wasteful (64 bytes is 4.6% of the total onion size...), I'm pretty sure we can do much better (for example by encrypting the metadata with a key derived from the `offerId` and our `nodeKey` instead of signing it). This is an area that other implementations have given a lot of thought about, can you send a mail to the lightning-dev mailing list detailing the current DB-less scheme you're proposing and asking others to share the schemes they plan on implementing?

This way we can gather knowledge from many people and find a better scheme. What's nice is that all of this is transient and short-lived anyway, so we can change our DB-less scheme by simply updating eclair without caring about backwards-compatibility with the old scheme.